File: 2022/05/part1.rb
#!/usr/bin/env ruby

input = ARGV[0] == "ex" ? "example" : "input"

it = IO.foreach(File.join(__dir__, input), chomp: true)
game = it.take_while { !_1.empty? }
instructions = it.drop_while { !_1.empty? }.drop(1)

# Parse Game

size = game.last.split.size
game = game[0...-1].map { |line|
  size.times.map { |i| line[i * 4 + 1]  } # 1 5 9
}.reverse.transpose.map { |col| col.reject { _1 == " " } }

# Parse Instructions

# intructions are [from, to] arrays of one move.
instructions = instructions.flat_map { |line|
  line[/move (\d+) from (\d+) to (\d+)/]
  [[$2.to_i - 1, $3.to_i - 1]] * $1.to_i
}

# Play Game

instructions.each do
  game[_2].push game[_1].pop
end

puts game.map(&:last).join

File: 2025/09/part1.py
#!/usr/bin/env python3

from sys import argv

# 50


def parse() -> list[tuple[int, int]]:
	rv = []
	with open(argv[1], 'r') as file:
		for line in file:
			a, b = line.split(',')
			rv.append((int(a), int(b)))
		return rv


def size(a: tuple[int, int], b: tuple[int, int]) -> int:
	"""
	>>> size((3, 2), (1, 1))
	6
	"""
	a1, a2 = a
	b1, b2 = b
	return (abs(b1 - a1) + 1) * (abs(b2 - a2) + 1)


if __name__ == '__main__':
	tiles = parse()
	largest = 0
	for i, t1 in enumerate(tiles):
		for t2 in tiles[i + 1 :]:
			area = size(t1, t2)
			if area > largest:
				largest = area
	print(largest)
